python/ray/util/client/worker.py

```diff
@@ -68,14 +78,58 @@ def __init__(self,
                 break
             except grpc.FutureTimeoutError:
                 if conn_attempts >= connection_retries:
-                    raise ConnectionError("ray client connection timeout")
-                logger.info(f"Couldn't connect in {timeout} seconds, retrying")
-                timeout = timeout + 5
-                if timeout > MAX_TIMEOUT_SEC:
-                    timeout = MAX_TIMEOUT_SEC
-
+                    raise ConnectionError(
+                        "ray client connection timeout: channel can't connect")
+                logger.debug(
+                    f"Couldn't connect in {timeout} seconds, retrying")
+                timeout = backoff(timeout)
+
+        # Wrap the channel with the RayletDriverStub, allowing for unary
+        # requests.
         self.server = ray_client_pb2_grpc.RayletDriverStub(self.channel)
 
+        # Now, retry until both ray has been initialized and the server is
+        # responding (perhaps from the other end of the proxy).
+
+        # Retry on 500 errors, which may happen if the endpoint can't service
+        # our request yet, as per a proxy use-case.
+        conn_attempts = 0
+        timeout = INITIAL_TIMEOUT_SEC
+        ray_ready = False
+        while conn_attempts < connection_retries + 1:
+            conn_attempts += 1
+            try:
+                ray_ready = self.is_initialized()
+                break
+            except grpc.RpcError as e:
+                if e.code() == grpc.StatusCode.UNAVAILABLE:
+                    if conn_attempts >= connection_retries:
+                        raise ConnectionError(
+                            "ray client connection timeout: gRPC unavailable")
+                    logger.info(
+                        f"Ray server unavailable, retrying in {timeout}s...")
+                    logger.debug(f"Received when checking init: {e.details()}")
+                    time.sleep(timeout)
+                    timeout = backoff(timeout)
+                else:
+                    raise e
+
+        # Retry until ray initialization on the server is also complete, which
+        # may take a second for lazily instantiated servers.
+        conn_attempts = 0
+        timeout = INITIAL_TIMEOUT_SEC
+        while not ray_ready:
+            if conn_attempts >= (connection_retries + 1):
+                raise ConnectionError(
+                    "ray client connection timeout: ray not initialized")
+            logger.info("Waiting for Ray to become ready on the server, "
+                        f", retry in {timeout}s...")
+            time.sleep(timeout)
+            timeout = backoff(timeout)
+            conn_attempts += 1
+            ray_ready = self.is_initialized()
```

Comment:
Do we need three different retry loops for the different stages? Why not have just a single one that tests this last condition?